src/target/riscv/riscv-013.c

```diff
@@ -1370,14 +1370,31 @@ static int register_read_progbuf(struct target *target, uint64_t *value,
 {
 	assert(target->state == TARGET_HALTED);
 
-	if (number >= GDB_REGNO_FPR0 && number <= GDB_REGNO_FPR31)
-		return fpr_read_progbuf(target, value, number);
-	else if (number >= GDB_REGNO_CSR0 && number <= GDB_REGNO_CSR4095)
-		return csr_read_progbuf(target, value, number);
+	int res;
+	if (number >= GDB_REGNO_FPR0 && number <= GDB_REGNO_FPR31) {
+		res = fpr_read_progbuf(target, value, number);
+	} else if (number >= GDB_REGNO_CSR0 && number <= GDB_REGNO_CSR4095) {
+		res = csr_read_progbuf(target, value, number);
+	} else {
+		LOG_TARGET_ERROR(target, "Unexpected read of %s via program buffer.",
+				riscv_reg_gdb_regno_name(target, number));
+		return ERROR_FAIL;
+	}
+	if (res != ERROR_OK)
+		return res;
 
-	LOG_TARGET_ERROR(target, "Unexpected read of %s via program buffer.",
-			riscv_reg_gdb_regno_name(target, number));
-	return ERROR_FAIL;
+	unsigned int size_bits = register_size(target, number);
+	unsigned int value_bits = sizeof(*value) * CHAR_BIT;
+	assert(size_bits <= value_bits);
+	if (size_bits == value_bits || *value >> size_bits == 0)
+		return ERROR_OK;
+
+	LOG_TARGET_WARNING(target, "Value read for register %s of %" PRIx64
+			" does not fit into the size of the register (%u bits)."
+			" This is a HW bug. Truncating the value to fit into the register.",
+			riscv_reg_gdb_regno_name(target, number), *value, size_bits);
```

Comment:
This is a copy of https://github.com/riscv-collab/riscv-openocd/pull/1233#discussion_r1975272948 by @JanMatCodasip:

If this HW problem occurs, it is a serious thing, and we better not trust the returned value at all. 

Instead of truncating the value, I would recommend to change the message level to ERROR and end the operation with an error code.

Few additional details:

- Added prefix `0x` in front of the hex number
- I'd suggest minor change of word order for readability
- `does not fit into the size` --> `exceeds the size` (shorter and more formal)

```suggestion
	LOG_TARGET_ERROR(target, "Value 0x%" PRIx64 " read from register %s"
			" exceeds the size of the register (%u bits). This is a HW bug. "
			*value, riscv_reg_gdb_regno_name(target, number), size_bits);
	return ERROR_FAIL;
```

Otherwise this change looks good.

Reply:
I agree that this should be an error. While we do accommodate some targets which aren't fully RISC-V debug compliant (notably GD32VF103), this is serious and we can't expect anything.